Examine file ctime when checking if files have changed.

cbane opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

What is the purpose of this change? What does it change?
--------------------------------------------------------

This makes restic compare file ctimes (in addition to all of the existing checks) when it is deciding if a file has been modified.

<!--
Describe the changes here, as detailed as needed.
-->

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Yes, closes #2179.

<!--
Link issues and relevant forum posts here.
-->

Checklist
---------

- [X] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [X] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [X] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [X] I have run `gofmt` on the code in all commits
- [X] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [X] I'm done, this Pull Request is ready for review

Commit:
Examine file ctime when checking if files have changed.

cbane commented:
It looks like this change is breaking #2205 (or at least its unit tests). How do we want these to interact? The only idea I've had so far is disabling the ctime check if the ignore inode flag is on, but I'm open to anything that makes sense.

d3zd3z commented:
The only idea I've had so far is disabling the ctime check if the ignore inode flag is on, but I'm open to anything that makes sense.

The `ctime` value really only makes sense on the exact same inode, so this seems to me like a reasonable way of handling this.

fd0 reviewed (changes requested):
Thanks for taking a stab at this! This is a change which may leave users wondering what's happening, so we need to be extra careful to address this in the changelog. As far as I understood, there are some situations in which restic did not read the contents of the file, but does now (file with different content, but same length and `mtime` and inode). So for people depending on this (there are always people who depend on any behavior, no matter how obscure) will see that restic reads and hashes more files than before. Also, when metadata is changed, the `ctime` is updated and the file is re-read.

In my opinion, being correct and re-reading the contents if in doubt is the better strategy, so I'm in favor for this. Maybe we should add a paragraph to the changelog that people should open an issue if this breaks anything for them, so we can decide if we need a flag for it?

We need to clarify the following things before it can be merged:
 * Is it a bug or an enhancement (important for the changelog entry)? I'm currently leaning towards the latter, but I'm open to arguments.
 * The code needs to be rebased on master, the conflict with the ignore-inode option needs to be resolved
 * When `--ignore-inode` is turned on, I think the `ctime` should not be considered so I agree with @d3zd3z

Commit:
Don't check ctime when ignoring inode.

Commit:
Use existing setTimestamp function for ctime test and improve error checking.

Commit:
Update changelog text, and add pull request link.

cbane commented:
OK, I rebased against master, and it now skips the ctime check when the `--ignore-inode` option is given. I also updated the text for the changelog entry.

fd0 commented on changelog/unreleased/issue-2179:
```diff
@@ -0,0 +1,15 @@
+file should be noticed, and the modified file will be backed up. The ctime check
+will be disabled if the --ignore-inode flag was given.
+
+If this change causes problems for you, please open an issue, and we can look in
```
I love the wording!

fd0 commented on internal/archiver/archiver_test.go:
```diff
@@ -576,6 +577,19 @@ func TestFileChanged(t *testing.T) {
 			},
 		},
+		{
+			Name: "new-content-same-timestamp",
```
Table-driven tests are awesome, easily extendable for contributors! :)

fd0 reviewed (approved):
LGTM, thanks!

DurvalMenezes commented:
Please excuse me, but should'nt we be giving more warning to users about this problem? Please see my rationale [here](https://github.com/restic/restic/issues/2179#issuecomment-495654982).
